File: show/software/source/linux/npu/userspace/xnpu_inspect.c
/* SPDX-License-Identifier: MIT */
#include "xnpu.h"

#include <errno.h>
#include <stdio.h>
#include <string.h>

static const char *task_name(uint32_t task)
{
	return task == XNPU_TASK_BBOX ? "bbox" : "classification";
}

int main(int argc, char **argv)
{
	struct xnpu_package package;
	char digest[65];
	const char *name;
	size_t name_length;
	size_t index;
	int result;

	if (argc != 2) {
		fprintf(stderr, "usage: xnpu-inspect MODEL.xnpu\n");
		return 2;
	}
	result = xnpu_package_open(&package, argv[1]);
	if (result) {
		fprintf(stderr, "%s: %s\n", argv[1], strerror(-result));
		return 1;
	}
	name = xnpu_package_name(&package, &name_length);
	xnpu_sha256_hex(package.info.package_sha256, digest);
	printf("name=%.*s model_id=%u task=%s hardware_abi=%u layers=%u\n",
	       (int)name_length, name, package.info.model_id,
	       task_name(package.info.task), package.info.hardware_abi,
	       package.info.layer_count);
	printf("input=%ux%ux%u mode=%u layout=%u dtype=%u bytes=%u\n",
	       package.info.input.width, package.info.input.height,
	       package.info.input.channels, package.info.input_mode,
	       package.info.input.layout, package.info.input.dtype,
	       package.info.input.bytes);
	printf("output=%ux%ux%u layout=%u dtype=%u bytes=%u\n",
	       package.info.output.width, package.info.output.height,
	       package.info.output.channels, package.info.output.layout,
	       package.info.output.dtype, package.info.output.bytes);
	printf("parameters=%u scratch=%u required_caps=0x%08x\n",
	       package.info.parameter_bytes, package.info.scratch_bytes,
	       package.info.required_caps);
	printf("package_sha256=%s\n", digest);
	for (index = 0; index < package.section_count; ++index)
		printf("section=%u flags=0x%x bytes=%zu crc32=0x%08x\n",
		       package.sections[index].type,
		       package.sections[index].flags,
		       package.sections[index].size,
		       package.sections[index].crc32);
	xnpu_package_close(&package);
	return 0;
}

File: show/software/source/chiplab/IP/NPU/scripts/xnpu_inspect.py
#!/usr/bin/env python3
"""Inspect and validate an XNPU deployment package."""

from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path

from xnpu_format import (
    SECTION_LABELS,
    SECTION_METADATA,
    SECTION_NAME,
    load_package,
)

TASKS = {1: "bbox", 2: "classification"}
MODES = {0: "frame", 1: "packed_preload"}
LAYOUTS = {0: "linear", 1: "nhwc", 2: "nchw"}
DTYPES = {0: "u8", 1: "s8", 2: "u16", 3: "s16", 4: "u32", 5: "s32"}


def describe(path: Path) -> dict:
    package = load_package(path)
    header = package.header
    result = {
        "path": str(path),
        "file_sha256": hashlib.sha256(package.data).hexdigest(),
        "package_sha256": package.sha256.hex(),
        "model_id": header["model_id"],
        "name": package.section(SECTION_NAME).decode(),
        "hardware_abi": header["hardware_abi"],
        "task": TASKS.get(header["task"], f"unknown:{header['task']}"),
        "layers": header["layer_count"],
        "input": {
            "mode": MODES.get(header["input_mode"], f"unknown:{header['input_mode']}"),
            "shape": [
                header["input_width"],
                header["input_height"],
                header["input_channels"],
            ],
            "layout": LAYOUTS.get(header["input_layout"], "unknown"),
            "dtype": DTYPES.get(header["input_dtype"], "unknown"),
            "bytes": header["input_bytes"],
        },
        "output": {
            "shape": [
                header["output_width"],
                header["output_height"],
                header["output_channels"],
            ],
            "layout": LAYOUTS.get(header["output_layout"], "unknown"),
            "dtype": DTYPES.get(header["output_dtype"], "unknown"),
            "bytes": header["output_bytes"],
        },
        "parameter_bytes": header["parameter_bytes"],
        "scratch_bytes": header["scratch_bytes"],
        "required_caps": f"0x{header['required_caps']:08x}",
        "sections": [
            {
                "type": section.type,
                "flags": section.flags,
                "offset": section.offset,
                "bytes": len(section.data),
                "crc32": f"0x{section.crc32:08x}",
            }
            for section in package.sections.values()
        ],
    }
    if SECTION_LABELS in package.sections:
        result["labels"] = package.section(SECTION_LABELS).decode().splitlines()
    if SECTION_METADATA in package.sections:
        result["metadata"] = json.loads(package.section(SECTION_METADATA))
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--json", action="store_true", help="emit canonical JSON")
    parser.add_argument("package", type=Path)
    args = parser.parse_args()
    info = describe(args.package)
    if args.json:
        print(json.dumps(info, sort_keys=True, separators=(",", ":"), ensure_ascii=False))
    else:
        print(f"name: {info['name']} (id={info['model_id']})")
        print(f"hardware ABI: {info['hardware_abi']}  task: {info['task']}")
        print(f"layers: {info['layers']}  required caps: {info['required_caps']}")
        print(f"input: {info['input']}")
        print(f"output: {info['output']}")
        print(
            f"parameters: {info['parameter_bytes']}  scratch: {info['scratch_bytes']}"
        )
        print(f"package SHA-256: {info['package_sha256']}")
        print(f"file SHA-256: {info['file_sha256']}")
        for section in info["sections"]:
            print(
                "section "
                f"{section['type']}: offset={section['offset']} "
                f"bytes={section['bytes']} crc32={section['crc32']}"
            )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
